File: Medii de proiectare si programare/C#/tema7/networking/ClientObjectWorker.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using model.domains;
using service;

namespace networking
{
    public sealed class ClientObjectWorker : IServiceObserver
    {
        private readonly TcpClient _connection;
        private readonly IFormatter _formatter;

        private readonly IService _server;
        private readonly NetworkStream _stream;
        private volatile bool _connected;

        public ClientObjectWorker(IService server, TcpClient connection)
        {
            _server = server;
            _connection = connection;
            try
            {
                _stream = connection.GetStream();
                _formatter = new BinaryFormatter();
                _connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void TicketsSold(Buyer buyers)
        {
            try
            {
                SendResponse(new NewSellTicketsResponse(buyers));
            }
            catch (Exception e)
            {
                throw new Exception("Sending error: " + e);
            }
        }

        public void Run()
        {
            while (_connected)
            {
                try
                {
                    var request = _formatter.Deserialize(_stream);
                    var response = HandleRequest((Request)request);
                    if (response != null) SendResponse(response);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }

            try
            {
                _stream.Close();
                _connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private void SendResponse(Response response)
        {
            lock (_stream)
            {
                _formatter.Serialize(_stream, response);
                _stream.Flush();
            }
        }

        private Response HandleRequest(Request request)
        {
            Response response = null;
            if (request is LogInRequest)
            {
                var logInRequest = (LogInRequest)request;
                var employeeAtOffice = logInRequest._employeeAtOffice;
                try
                {
                    lock (_server)
                    {
                        _server.LogIn(employeeAtOffice.Email, employeeAtOffice.Password, this);
                    }

                    return new OkResponse();
                }
                catch (Exception e)
                {
                    _connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is LogOutRequest)
            {
                var logOutRequest = (LogOutRequest)request;
                var employeeAtOffice = logOutRequest.EmployeeAtOffice;
                try
                {
                    lock (_server)
                    {
                        _server.LogOut(employeeAtOffice, this);
                    }

                    _connected = false;
                    return new OkResponse();
                }
                catch (Exception e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is SellTicketsRequest)
            {
                var sellTicketsRequest = (SellTicketsRequest)request;
                var buyer = sellTicketsRequest.Buyer;
                try
                {
                    lock (_server)
                    {
                        _server.SellTickets(buyer.ShowName1, buyer.BuyerName1, buyer.NoOfTicketsBought1);
                    }

                    return new OkResponse();
                }
                catch (Exception e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is GetAllShowRequest)
            {
                var getAllShowRequest = (GetAllShowRequest)request;
                var employeeAtOffice = getAllShowRequest.EmployeeAtOffice;
                try
                {
                    List<Show> list;
                    lock (_server)
                    {
                        list = _server.GetAllShows(employeeAtOffice);
                    }

                    return new GetAllShowsResponse(list);
                }
                catch (Exception e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            return response;
        }
    }
}

File: Medii de proiectare si programare/java/tema7/temaJavaMPP/ChatNetworking/src/main/java/com/example/objectProtocol/ClientObjectWorker.java
package com.example.objectProtocol;

import com.example.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;


public class ClientObjectWorker implements Runnable, IServiceObserver {

    private final IService server;
    private final Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream outputStream;

    private volatile boolean connected;

    public ClientObjectWorker(IService service, Socket connection) {
        this.server = service;
        this.connection = connection;
        try {
            outputStream = new ObjectOutputStream(connection.getOutputStream());
            outputStream.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while (connected) {
            try {
                Object request = input.readObject();
                Object response = handleRequest((Request) request);
                if (response != null) {
                    sendResponse((Response) response);
                }
            } catch (IOException | ClassNotFoundException e) {
                connected = false;
            }
        }
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        try {
            input.close();
            outputStream.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("ERROR" + e);
        }
    }

    private void sendResponse(Response response) throws IOException {
        synchronized (outputStream) {
            System.out.println("sendResponse Object");
            outputStream.writeObject(response);
            outputStream.flush();
        }
    }

    private Object handleRequest(Request request) {
        Response response = null;
        if (request instanceof LogInRequest) {
            LogInRequest logInRequest = (LogInRequest) request;
            EmployeeAtOffice employeeAtOffice = logInRequest.getEmployeeAtOffice();
            try {
                server.logInEmployee(employeeAtOffice.getEmail(), employeeAtOffice.getPassword(), this);
                return new OKResponse();
            } catch (Exception e) {
                connected = false;
                return new ErrorResponse(e.getMessage());
            }
        }

        if (request instanceof LogOutRequest) {
            LogOutRequest logOutRequest = (LogOutRequest) request;
            EmployeeAtOffice employeeAtOffice = logOutRequest.getEmployeeAtOffice();
            try {
                server.logOut(employeeAtOffice);
                connected = false;
                return new OKResponse();
            } catch (Exception e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if (request instanceof SellTicketsRequest) {
            SellTicketsRequest sellTicketsRequest = (SellTicketsRequest) request;
            Buyers buyers = sellTicketsRequest.getBuyers();
            try {
                server.sellTickets(buyers.getShowName(), buyers.getBuyerName(), buyers.getNoTicketsBought());
                return new OKResponse();
            } catch (Exception e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof UpdateShowRequest) {
            UpdateShowRequest updateShowRequest = (UpdateShowRequest) request;
            Show show = updateShowRequest.getShow();
            try {
                server.updateShow(show);
                return new OKResponse();
            } catch (Exception e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof GetAllShowsRequest) {
            System.out.println();
            GetAllShowsRequest getAllShowsRequest = (GetAllShowsRequest) request;
            EmployeeAtOffice employeeAtOffice = getAllShowsRequest.getEmployeeAtOffice();
            try {
                List<Show> list = server.getAllShows(employeeAtOffice);
                return new GetShowsResponse(list);
            } catch (Exception e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        return response;
    }

    public void ticketSold(Buyers buyers) throws Exception {
        try {
            System.out.println("tivketSold ObjectWorker");
            sendResponse(new SellTicketResponse(buyers));
        } catch (IOException e) {
            throw new Exception(e);
        }
    }
}
